### PR TITLE
Enable MiMa

### LINKED ISSUE #190: Enable MiMa
Follow-up to https://github.com/typelevel/munit-cats-effect/pull/182#issuecomment-1024114238.

### PR DESCRIPTION
**Pull request by danicheg**
Fixes #190.
There are many different binary compatibility issues, so I suggest just starting checking binary compatibility from the `1.0.6` version (the last released is `1.0.7`). Any concerns about this?

### CODE REVIEW

**Comment by armanbilge** [comment]
What are the MiMa issues?

We can usually ignore MiMa issues on Scala 3 since it doesn't automatically recognized package-privates.

File: build.sbt [hunk 1]
```diff
@@ -21,7 +21,12 @@ lazy val ce3 = crossProject(JSPlatform, JVMPlatform)
       "org.scalameta" %%% "munit" % "0.7.29",
       "org.typelevel" %%% "cats-effect" % "3.3.5"
     ),
-    mimaPreviousArtifacts := Set.empty
+    // we are checking binary compatibility from the 1.0.6 version
+    mimaPreviousArtifacts ~= {
+      _.filter { m =>
+        VersionNumber(m.revision).matchesSemVer(SemanticSelector(">=1.0.6"))
+      }
+    }
```
Comment (armanbilge) [inline build.sbt:29]: Alternately:
```scala
ThisBuild / tlVersionIntroduced := List("2.12", "2.13", "3").map(_ -> "1.0.6").toMap
```

**Comment by danicheg** [comment]
@armanbilge It founds out that `munit-cats-effect` has binary issues with almost every version from `1.0.0` to `1.0.5`. If you wonder about particular problems, here is the small part of it:
```
[error] munit-cats-effect-2: Failed binary compatibility check against org.typelevel:munit-cats-effect-2_2.13:1.0.0 (e:info.versionScheme=early-semver)! Found 2 potential problems
[error]  * object munit.CatsEffectFixtures#ResourceSuiteLocalFixture does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("munit.CatsEffectFixtures$ResourceSuiteLocalFixture$")
[error]  * method ResourceSuiteLocalFixture()munit.CatsEffectFixtures#ResourceSuiteLocalFixture# in class munit.CatsEffectSuite has a different result type in current version, where it is munit.CatsEffectFixturesPlatform#ResourceSuiteLocalFixture# rather than munit.CatsEffectFixtures#ResourceSuiteLocalFixture#
[error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("munit.CatsEffectSuite.ResourceSuiteLocalFixture")
[error] (ce3JS / mimaPreviousClassfiles) sbt.librarymanagement.ResolveException: Error downloading org.typelevel:munit-cats-effect-3_sjs1_2.13:1.0.4
[error]   Not found
[error]   not found: /home/danicheg/.ivy2/local/org.typelevel/munit-cats-effect-3_sjs1_2.13/1.0.4/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/org/typelevel/munit-cats-effect-3_sjs1_2.13/1.0.4/munit-cats-effect-3_sjs1_2.13-1.0.4.pom
[error] (ce3JVM / mimaPreviousClassfiles) sbt.librarymanagement.ResolveException: Error downloading org.typelevel:munit-cats-effect-3_2.13:1.0.4
evel:munit-cats-effect-2_2.13:1.0.2 (e:info.versionScheme=early-semver)! Found 3 potential problems
[error]  * object munit.CatsEffectFixtures#ResourceSuiteLocalFixture does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("munit.CatsEffectFixtures$ResourceSuiteLocalFixture$")
[error]  * class munit.CatsEffectFixtures#ResourceSuiteLocalFixture#FixtureNotInstantiatedException does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("munit.CatsEffectFixtures$ResourceSuiteLocalFixture$FixtureNotInstantiatedException")
[error]  * method ResourceSuiteLocalFixture()munit.CatsEffectFixtures#ResourceSuiteLocalFixture# in class munit.CatsEffectSuite has a different result type in current version, where it is munit.CatsEffectFixturesPlatform#ResourceSuiteLocalFixture# rather than munit.CatsEffectFixtures#ResourceSuiteLocalFixture#
[error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("munit.CatsEffectSuite.ResourceSuiteLocalFixture")
[error] munit-cats-effect-3: Failed binary compatibility check against org.typelevel:munit-cats-effect-3_sjs1_2.13:1.0.5 (e:info.versionScheme=early-semver)! Found 1 potential problems
[error]  * method ioRuntime()cats.effect.unsafe.IORuntime in class munit.CatsEffectSuite does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("munit.CatsEffectSuite.ioRuntime")
[error] munit-cats-effect-3: Failed binary compatibility check against org.typelevel:munit-cats-effect-3_2.13:1.0.1 (e:info.versionScheme=early-semver)! Found 3 potential problems
[error]  * object munit.CatsEffectFixtures#ResourceSuiteLocalFixture does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("munit.CatsEffectFixtures$ResourceSuiteLocalFixture$")
[error]  * method ResourceSuiteLocalFixture()munit.CatsEffectFixtures#ResourceSuiteLocalFixture# in class munit.CatsEffectSuite has a different result type in current version, where it is munit.CatsEffectFixturesPlatform#ResourceSuiteLocalFixture# rather than munit.CatsEffectFixtures#ResourceSuiteLocalFixture#
[error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("munit.CatsEffectSuite.ResourceSuiteLocalFixture")
[error]  * method ioRuntime()cats.effect.unsafe.IORuntime in class munit.CatsEffectSuite does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("munit.CatsEffectSuite.ioRuntime")
[info] Fetched artifacts of 
[error] munit-cats-effect-2: Failed binary compatibility check against org.typelevel:munit-cats-effect-2_2.13:1.0.2 (e:info.versionScheme=early-semver)! Found 3 potential problems
[error]  * object munit.CatsEffectFixtures#ResourceSuiteLocalFixture does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("munit.CatsEffectFixtures$ResourceSuiteLocalFixture$")
[error]  * class munit.CatsEffectFixtures#ResourceSuiteLocalFixture#FixtureNotInstantiatedException does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("munit.CatsEffectFixtures$ResourceSuiteLocalFixture$FixtureNotInstantiatedException")
[error]  * method ResourceSuiteLocalFixture()munit.CatsEffectFixtures#ResourceSuiteLocalFixture# in class munit.CatsEffectSuite has a different result type in current version, where it is munit.CatsEffectFixturesPlatform#ResourceSuiteLocalFixture# rather than munit.CatsEffectFixtures#ResourceSuiteLocalFixture#
[error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("munit.CatsEffectSuite.ResourceSuiteLocalFixture")
```
I'm sure that we shouldn't add filters for those problems because we are enabling MiMa only from `1.0.7`. From my perspective, it would be fine even if we'll start checking it from the current version, not from the previous.

**Comment by armanbilge** [comment]
Yes, I agree we should not filter these issues. I was curious if there is any hope to fix them.

Thanks for posting that. Actually, now I see those problems seem to be my fault in https://github.com/typelevel/munit-cats-effect/pull/104 😅 

I'm not hopeful, but maybe it can be fixed.

**Comment by danicheg** [comment]
@armanbilge if I get it right, if we fix it, that would mean we will have binary issues with `1.0.6` and `1.0.7` versions 😄

**Comment by armanbilge** [comment]
No, I think we may be able to fix it for everything :)

**Comment by danicheg** [comment]
Yeah, in theory, I may guess it so. But can we do it separately from this PR, WDYT?

**Comment by armanbilge** [comment]
Nope, I was wrong. There's at least one method that is completely incompatible between 1.0.0-1.0.5 and 1.0.6-1.0.7. So somebody has to break.

```
[error]  * method ResourceSuiteLocalFixture()munit.CatsEffectFixturesPlatform#ResourceSuiteLocalFixture# in class munit.CatsEffectSuite has a different result type in current version, where it is munit.CatsEffectFixtures#ResourceSuiteLocalFixture# rather than munit.CatsEffectFixturesPlatform#ResourceSuiteLocalFixture#
```

👍 from me, this PR seems the right way forward.

**Comment by danicheg** [comment]
So, let's merge this and see what happens. If someone will have concerns about it - feel free to reconfigure MiMa as needed.

**Pull request merged**